test/IntegrationTests/MSTest.Acceptance.IntegrationTests/ThreadContextTests.cs

```diff
@@ -117,12 +209,562 @@ public void TestMethod1()
+        <AssemblyCleanupTimeout>10007</AssemblyCleanupTimeout>
+    </MSTest>
+</RunSettings>
+
+#file CultureFlowsThreadContextProject.csproj
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <EnableMSTestRunner>true</EnableMSTestRunner>
+    <TargetFrameworks>$TargetFrameworks$</TargetFrameworks>
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="MSTest.TestAdapter" Version="$MSTestVersion$" />
+    <PackageReference Include="MSTest.TestFramework" Version="$MSTestVersion$" />
+    <PackageReference Include="Microsoft.Testing.Platform" Version="$MicrosoftTestingPlatformVersion$" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="*.runsettings">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
+
+</Project>
+
+#file UnitTest1.cs
+namespace CultureFlowsThreadContextProject;
+
+using System;
+using System.Globalization;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+[TestClass]
+public class UnitTest1
+{
+    private const string AssemblyInitCultureCodeName = "th-TH";
+    private const string ClassInitCultureCodeName = "tr-TR";
+    private const string TestInitCultureCodeName = "sv-SE";
+    private const string TestMethodCultureCodeName = "ak-GH";
+    private const string TestCleanupCultureCodeName = "pt-BR";
+    private const string ClassCleanupCultureCodeName = "hu-HU";
+
+    [AssemblyInitialize]
+    public static void AssemblyInitialize(TestContext context)
+    {
+        CultureInfo.CurrentCulture = new CultureInfo(AssemblyInitCultureCodeName);
+    }
+
+    [ClassInitialize]
+    public static void ClassInitialize(TestContext context)
+    {
+        Assert.AreEqual(AssemblyInitCultureCodeName, CultureInfo.CurrentCulture.Name, "ClassInitialize culture check failed");
+        CultureInfo.CurrentCulture = new CultureInfo(ClassInitCultureCodeName);
+    }
+
+    [TestInitialize]
+    public void TestInitialize()
+    {
+        Assert.AreEqual(ClassInitCultureCodeName, CultureInfo.CurrentCulture.Name, "TestInitialize culture check failed");
+        CultureInfo.CurrentCulture = new CultureInfo(TestInitCultureCodeName);
+    }
+
+    [TestMethod]
+    public void TestMethod1()
+    {
+        Assert.AreEqual(TestInitCultureCodeName, CultureInfo.CurrentCulture.Name, "TestMethod culture check failed");
+        CultureInfo.CurrentCulture = new CultureInfo(TestMethodCultureCodeName);
+    }
+
+    [TestCleanup]
+    public void TestCleanup()
+    {
+        Assert.AreEqual(TestMethodCultureCodeName, CultureInfo.CurrentCulture.Name, "TestCleanup culture check failed");
+        CultureInfo.CurrentCulture = new CultureInfo(TestCleanupCultureCodeName);
+    }
+
+    [ClassCleanup]
+    public static void ClassCleanup()
+    {
+        if (Environment.GetEnvironmentVariable("MSTEST_TEST_FLOW_CONTEXT") == "true")
```

Comment:
All the cases where you will see the usage of this env var are cases where I would need input from reviewers. Basically we can either make the code to work like the "sync" case with the same "inconsistencies" to ensure everything is simple. Or we can try to consider that we want to bring some kind of functional logic to the way we save and restore the execution context (what is currently done).

Reply:
In this case, is the meaning of the flag: MSTEST restore class init context in class cleanup? 

I think that is reasonable behavior in some cases, and not in others. We don't know how many tests will run from the current class, and we don't know what they do or if they even complete, so ensuring that the state is the same as it was when ClassInit completed seems reasonable. 

On the other hand, if I think about storing other data, e.g. in thread local storage, or if we added new functionality that would have explicit ClassContext where users would be able to add their data, then I would like the data from tests to aggregate in the class context, e.g. adding all the database table names my Ordered tests created, and then cleaning them up in ClassCleanup, so I don't have to have a separate DB for each test.

Sounds kinda wild, but if this happens it would be nice to have consistent behavior.

Reply:
> In this case, is the meaning of the flag: MSTEST restore class init context in class cleanup?

For this exact case of class cleanup, the current non-threaded behavior is to use whatever state left by the last executing test. My implementation is trying to work similarly to async/await with a `using` (try/finally) principle, so the class cleanup will reuse the state left by the init but not by what's under it.

> I think that is reasonable behavior in some cases, and not in others.

Agreed, that's why I am calling it out :) I am on the fence about keeping current logic that seems logical (to me and seems to you) or to ensure we have always the same (or as close as possible) to not having the timeout. My main cons about having the same behavior is that anyways when you have sync vs async you do have differences by runtime and you have to know it so it seems reasonable to also know that having non-cooperative timeout changes the execution model. Obviously that means documenting it out correctly.

As you, and other reviewers can see, the problem only happen for cleanup states. And the reported bug was only complaining about inits not flowing down to the test method.